cli/flox-rust-sdk/src/providers/build.rs

```diff
@@ -161,3 +161,435 @@ fn read_output_to_channel(
+        let file_name = String::from("bar");
+        let file_content = String::from("some content");
+
+        let manifest = formatdoc! {r#"
+            version = 1
+
+            [build.{package_name}]
+            command = """
+                mkdir $out
+                cp {file_name} $out/{file_name}
+            """
+        "#};
+
+        let (flox, _temp_dir_handle) = flox_instance();
+        let mut env = new_path_environment(&flox, &manifest);
+        let env_path = env.parent_path().unwrap();
+
+        // This file is not accessible from a pure build.
+        fs::write(env_path.join(&file_name), &file_content).unwrap();
+        assert_build_status(&flox, &mut env, &package_name, false);
+
+        let dir = result_dir(&env_path, &package_name);
+        assert!(!dir.exists());
+    }
+
+    #[test]
+    fn build_sandbox_off() {
+        let package_name = String::from("foo");
+        let file_name = String::from("bar");
+        let file_content = String::from("some content");
+
+        let manifest = formatdoc! {r#"
+            version = 1
+
+            [build.{package_name}]
+            sandbox = "off"
+            command = """
+                mkdir $out
+                cp {file_name} $out/{file_name}
+            """
+        "#};
+
+        let (flox, _temp_dir_handle) = flox_instance();
+        let mut env = new_path_environment(&flox, &manifest);
+        let env_path = env.parent_path().unwrap();
+
+        // This file is accessible from an impure build.
+        fs::write(env_path.join(&file_name), &file_content).unwrap();
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+    }
+
+    #[test]
+    #[ignore] // TODO: Intermittently uses cache.
+    fn build_cache_sandbox_pure_uses_cache() {
+        let package_name = String::from("foo");
+        let file_name = String::from("bar");
+        let file_content = String::from("some content");
+
+        let manifest = formatdoc! {r#"
+            version = 1
+
+            [build.{package_name}]
+            sandbox = "pure"
+            command = """
+                mkdir $out
+                echo -n {file_content} > $out/{file_name}
+            """
+        "#};
+
+        let (flox, _temp_dir_handle) = flox_instance();
+        let mut env = new_path_environment(&flox, &manifest);
+        let env_path = env.parent_path().unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_first = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_second = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        assert_eq!(dir_second, dir_first);
+    }
```

Comment:
The second build uses a different cache file causing the input addressed result store path to differ:

**first build**

```
stdout: this derivation will be built:
stdout:   /nix/store/xkihqlkarsn69vwy31fqp5q76irn83pl-foo-0.0.0.drv
stdout: building '/nix/store/xkihqlkarsn69vwy31fqp5q76irn83pl-foo-0.0.0.drv'...
stdout: foo> ---
stdout: foo> Input checksums:
stdout: foo> 1276481102f218c981e0324180bafd9f  /nix/store/8v6kxjpm58b7912ylwvir304nk4dxh0j-e0996333-foo-src.tar
stdout: foo> 3b7230fcc99c89b8728753b8875b05fa  /nix/store/8p4750879rj74w1ybk9zcsi54akq2lpl-e0996333-foo-build.bash
stdout: foo> 1025e9248c75a7a96a011ea8de412651  /nix/store/zxkc1m42zpbjl3i51cy6b2jqjwip0pv4-e0996333-foo-buildCache.tar
stdout: foo> ---
```

**second build**
```
stdout: this derivation will be built:
stdout:   /nix/store/i9d70n5mrkxs80gm9mysmw7hklgfms66-foo-0.0.0.drv
stdout: building '/nix/store/i9d70n5mrkxs80gm9mysmw7hklgfms66-foo-0.0.0.drv'...
stdout: foo> ---
stdout: foo> Input checksums:
stdout: foo> 1276481102f218c981e0324180bafd9f  /nix/store/8v6kxjpm58b7912ylwvir304nk4dxh0j-e0996333-foo-src.tar
stdout: foo> 3b7230fcc99c89b8728753b8875b05fa  /nix/store/8p4750879rj74w1ybk9zcsi54akq2lpl-e0996333-foo-build.bash
stdout: foo> 123730362fad216a028e0f73e3e77d43  /nix/store/1cwb4hrjpnfslvavvikh1zh5ncl0qn4j-e0996333-foo-buildCache.tar
stdout: foo> ---
```


<details>
```
stdout: make: Entering directory '/private/tmp/.tmpqFsmuv/temp/.tmpfHvM1s'
stdout: /nix/store/yj6m46w8yzc7missw9f4yfmvx1ai9c4q-gnutar-1.35/bin/tar -cf - --no-recursion -T <(/nix/store/45kxi48hp60a6gvammiihx246kspmr7i-git-minimal-2.45.2/bin/git ls-files) > /tmp/nix-shell.1fbs3B/e0996333-foo-src.tar
stderr: fatal: not a git repository (or any of the parent directories): .git
stdout: /nix/store/pldbvp0p2z2pa76h17g797vvk201155r-coreutils-9.5/bin/rm -f /tmp/nix-shell.1fbs3B/e0996333-foo-buildCache.tar
stdout: Rendering foo build script to /tmp/nix-shell.1fbs3B/e0996333-foo-build.bash
stdout: Building foo-0.0.0 in Nix sandbox (pure) mode
stdout: set -o pipefail && /nix/store/0gsq84l1imlir3xlqd6s16lgxlbq6fnf-nix-2.18.5/bin/nix --extra-experimental-features "flakes nix-command" build -L --file /nix/store/5yi0l4c512jjd4zbdp4xnsxzcql9g6i5-package-builder-1.0.0/libexec/build-manifest.nix --argstr name "foo-0.0.0" --argstr srcTarball "/tmp/nix-shell.1fbs3B/e0996333-foo-src.tar" --argstr flox-env "/private/tmp/.tmpqFsmuv/temp/.tmpfHvM1s/.flox/run/aarch64-darwin.name" --argstr install-prefix "/tmp/store_e09963338dff5b0761618ec358260d5f-foo-0.0.0"  --argstr buildScript "/tmp/nix-shell.1fbs3B/e0996333-foo-build.bash" --argstr buildCache "/tmp/nix-shell.1fbs3B/e0996333-foo-buildCache.tar" --out-link "result-foo" '^*' 2>&1 | tee /tmp/nix-shell.1fbs3B/tmp.fAARb0EA4F-build-foo.log
stdout: this derivation will be built:
stdout:   /nix/store/xkihqlkarsn69vwy31fqp5q76irn83pl-foo-0.0.0.drv
stdout: building '/nix/store/xkihqlkarsn69vwy31fqp5q76irn83pl-foo-0.0.0.drv'...
stdout: foo> ---
stdout: foo> Input checksums:
stdout: foo> 1276481102f218c981e0324180bafd9f  /nix/store/8v6kxjpm58b7912ylwvir304nk4dxh0j-e0996333-foo-src.tar
stdout: foo> 3b7230fcc99c89b8728753b8875b05fa  /nix/store/8p4750879rj74w1ybk9zcsi54akq2lpl-e0996333-foo-build.bash
stdout: foo> 1025e9248c75a7a96a011ea8de412651  /nix/store/zxkc1m42zpbjl3i51cy6b2jqjwip0pv4-e0996333-foo-buildCache.tar
stdout: foo> ---
stdout: rm /tmp/nix-shell.1fbs3B/e0996333-foo-build.bash
stdout: make: Leaving directory '/private/tmp/.tmpqFsmuv/temp/.tmpfHvM1s'
stdout: make: Entering directory '/private/tmp/.tmpqFsmuv/temp/.tmpfHvM1s'
stdout: /nix/store/yj6m46w8yzc7missw9f4yfmvx1ai9c4q-gnutar-1.35/bin/tar -cf - --no-recursion -T <(/nix/store/45kxi48hp60a6gvammiihx246kspmr7i-git-minimal-2.45.2/bin/git ls-files) > /tmp/nix-shell.1fbs3B/e0996333-foo-src.tar
stderr: fatal: not a git repository (or any of the parent directories): .git
stdout: /nix/store/pldbvp0p2z2pa76h17g797vvk201155r-coreutils-9.5/bin/rm -f /tmp/nix-shell.1fbs3B/e0996333-foo-buildCache.tar
stdout: Rendering foo build script to /tmp/nix-shell.1fbs3B/e0996333-foo-build.bash
stdout: Building foo-0.0.0 in Nix sandbox (pure) mode
stdout: set -o pipefail && /nix/store/0gsq84l1imlir3xlqd6s16lgxlbq6fnf-nix-2.18.5/bin/nix --extra-experimental-features "flakes nix-command" build -L --file /nix/store/5yi0l4c512jjd4zbdp4xnsxzcql9g6i5-package-builder-1.0.0/libexec/build-manifest.nix --argstr name "foo-0.0.0" --argstr srcTarball "/tmp/nix-shell.1fbs3B/e0996333-foo-src.tar" --argstr flox-env "/private/tmp/.tmpqFsmuv/temp/.tmpfHvM1s/.flox/run/aarch64-darwin.name" --argstr install-prefix "/tmp/store_e09963338dff5b0761618ec358260d5f-foo-0.0.0"  --argstr buildScript "/tmp/nix-shell.1fbs3B/e0996333-foo-build.bash" --argstr buildCache "/tmp/nix-shell.1fbs3B/e0996333-foo-buildCache.tar" --out-link "result-foo" '^*' 2>&1 | tee /tmp/nix-shell.1fbs3B/tmp.ar3WDLpUl2-build-foo.log
stdout: this derivation will be built:
stdout:   /nix/store/i9d70n5mrkxs80gm9mysmw7hklgfms66-foo-0.0.0.drv
stdout: building '/nix/store/i9d70n5mrkxs80gm9mysmw7hklgfms66-foo-0.0.0.drv'...
stdout: foo> ---
stdout: foo> Input checksums:
stdout: foo> 1276481102f218c981e0324180bafd9f  /nix/store/8v6kxjpm58b7912ylwvir304nk4dxh0j-e0996333-foo-src.tar
stdout: foo> 3b7230fcc99c89b8728753b8875b05fa  /nix/store/8p4750879rj74w1ybk9zcsi54akq2lpl-e0996333-foo-build.bash
stdout: foo> 123730362fad216a028e0f73e3e77d43  /nix/store/1cwb4hrjpnfslvavvikh1zh5ncl0qn4j-e0996333-foo-buildCache.tar
stdout: foo> ---
stdout: rm /tmp/nix-shell.1fbs3B/e0996333-foo-build.bash
stdout: make: Leaving directory '/private/tmp/.tmpqFsmuv/temp/.tmpfHvM1s'
thread 'providers::build::tests::build_cache_sandbox_pure_uses_cache' panicked at flox-rust-sdk/src/providers/build.rs:406:9:
assertion `left == right` failed
  left: "/nix/store/j0s1xxzyx1kn2lhd5rb9cp23nkv0d19a-foo-0.0.0"
 right: "/nix/store/f758132viqyk6jv5ryywnkpsi5ypfmj5-foo-0.0.0"
```

</details>

Reply:
Glad you found the input checksums useful! I added that as part of my own debugging efforts to detect when I was and was not expecting a pure build to be performed.

> The second build uses a different cache file causing the input addressed result store path to differ:

Yes ... that makes sense. Typically that second build doesn't update the build cache so when you perform the third build Nix will know it does not need to perform it.

Not sure if there's a question here?

Reply:
Like in the previous one I think we assume that store paths won't change, but store paths _do_ change maybe even though the cache is working.

Are you saying we have to build three packages and compare the path of build 2 == 3 for that assumption to be valid?
Should we assert the cache being used differently (e.g a similar random string being initialized once like above?)

Reply:
> Are you saying we have to build three packages and compare the path of build 2 == 3 for that assumption to be valid?

Essentially yes:
1. first build starts with an empty build cache, finishes by populating a build cache with object files
    * --> Changes buildCache input as a result
2. second build starts with a populated build cache, and does not need to change it
    * --> That build is fast, but Nix performs the build all the same because the buildCache input changed
3. third build starts with the same build cache as build 2
    * --> Nix spots that no inputs have changed so it doesn't bother kicking off the build at all

Reply:
Makes sense yeah the input needs to be there before you build and we don't know what the input is supposed to be without building at least once